File: sdk/go/auth/auth_manager.go
// Package auth provides authentication utilities for the Solid Sidecar Go SDK.
//
// Phase 27 - SDK/Client Compatibility Layer
// Status: STABLE - Production Ready
package auth

import (
	"context"
	"crypto/tls"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/outlaw-dame/solid-sidecar/sdk/go/pkg/types"
)

// ErrAuthenticationFailed represents an authentication failure
var ErrAuthenticationFailed = errors.New("authentication failed")

// ErrTokenExpired represents an expired token
var ErrTokenExpired = errors.New("token expired")

// ErrNoToken represents missing token
var ErrNoToken = errors.New("no token available")

// ErrInvalidIssuer represents an invalid issuer
var ErrInvalidIssuer = errors.New("invalid issuer")

// AuthManager provides OAuth2/DPoP authentication management.
type AuthManager struct {
	mu sync.RWMutex

	// tokenManager manages access tokens
	tokenManager *TokenManager

	// keyStore manages DPoP keys
	keyStore *DPoPKeyStore

	// httpClient is the underlying HTTP client for token exchange
	httpClient *http.Client

	// issuer is the OAuth2 issuer URL
	issuer string

	// clientID is the OAuth2 client ID
	clientID string

	// clientSecret is the OAuth2 client secret (optional for confidential clients)
	clientSecret string

	// redirectURI is the OAuth2 redirect URI
	redirectURI string

	// scope is the requested scope
	scope string

	// currentKeyID is the current DPoP key ID
	currentKeyID string

	// autoRefresh indicates if tokens should be auto-refreshed
	autoRefresh bool

	// onTokenRefresh is called when a token is refreshed
	onTokenRefresh func(token string)
}

// AuthManagerOptions contains options for creating an AuthManager.
type AuthManagerOptions struct {
	// Issuer is the OAuth2 issuer URL
	Issuer string

	// ClientID is the OAuth2 client ID
	ClientID string

	// ClientSecret is the OAuth2 client secret
	ClientSecret string

	// RedirectURI is the OAuth2 redirect URI
	RedirectURI string

	// Scope is the requested scope
	Scope string

	// KeyStore is an existing DPoP key store (optional)
	KeyStore *DPoPKeyStore

	// TokenManager is an existing token manager (optional)
	TokenManager *TokenManager

	// HTTPClient is an existing HTTP client (optional)
	HTTPClient *http.Client

	// AutoRefresh indicates if tokens should be auto-refreshed
	AutoRefresh bool
}

// NewAuthManager creates a new AuthManager.
//
// Parameters:
//   - options: AuthManager options
//
// Returns:
//   - A new AuthManager instance
//   - Error if options are invalid
func NewAuthManager(options *AuthManagerOptions) (*AuthManager, error) {
	if options == nil {
		options = &AuthManagerOptions{}
	}

	// Validate issuer
	if options.Issuer == "" {
		return nil, ErrInvalidIssuer
	}

	// Parse issuer URL
	parsedIssuer, err := url.Parse(options.Issuer)
	if err != nil {
		return nil, fmt.Errorf("%w: invalid issuer URL: %v", ErrInvalidIssuer, err)
	}

	if parsedIssuer.Scheme == "" {
		parsedIssuer.Scheme = "https"
	}

	// Create HTTP client if not provided
	httpClient := options.HTTPClient
	if httpClient == nil {
		transport := &http.Transport{
			IdleConnTimeout:       90 * time.Second,
			ResponseHeaderTimeout: 10 * time.Second,
			MaxIdleConns:          100,
			MaxIdleConnsPerHost:   10,
		}

		// Allow insecure for localhost
		if strings.Contains(parsedIssuer.Host, "localhost") || strings.Contains(parsedIssuer.Host, "127.0.0.1") {
			transport.TLSClientConfig = &tls.Config{
				InsecureSkipVerify: true,
			}
		} else {
			transport.TLSClientConfig = &tls.Config{
				MinVersion: tls.VersionTLS12,
			}
		}

		httpClient = &http.Client{
			Transport: transport,
			Timeout:   30 * time.Second,
		}
	}

	// Create key store if not provided
	keyStore := options.KeyStore
	if keyStore == nil {
		keyStore = NewDPoPKeyStore(&DPoPKeyStoreOptions{
			Algorithm: RS256,
			KeySize:   2048,
		})
	}

	// Create token manager if not provided
	tokenManager := options.TokenManager
	if tokenManager == nil {
		tokenManager = NewTokenManager()
	}

	return &AuthManager{
		tokenManager: tokenManager,
		keyStore:     keyStore,
		httpClient:   httpClient,
		issuer:       parsedIssuer.String(),
		clientID:     options.ClientID,
		clientSecret: options.ClientSecret,
		redirectURI:  options.RedirectURI,
		scope:        options.Scope,
		autoRefresh:  options.AutoRefresh,
	}, nil
}

// SetToken sets the access token directly.
//
// Parameters:
//   - token: The access token
//   - expiresIn: The lifetime in seconds
func (am *AuthManager) SetToken(token string, expiresIn int64) {
	am.tokenManager.SetToken(&types.TokenResponse{
		AccessToken: token,
		TokenType:   "DPoP",
		ExpiresIn:   expiresIn,
		IssuedAt:    time.Now().UTC(),
	}, expiresIn)
}

// GetAccessToken returns the current access token.
// If auto-refresh is enabled and the token is expired, it will attempt to refresh.
//
// Returns:
//   - The access token
//   - Error if no token or refresh fails
func (am *AuthManager) GetAccessToken() (string, error) {
	am.mu.Lock()
	defer am.mu.Unlock()

	// Check if token exists and is valid
	if token := am.tokenManager.GetAccessToken(); token != "" {
		if !am.tokenManager.IsExpired() {
			return token, nil
		}

		// Token is expired - try to refresh if auto-refresh is enabled
		if am.autoRefresh && am.clientSecret != "" {
			if err := am.refreshToken(); err == nil {
				return am.tokenManager.GetAccessToken(), nil
			}
		}
	}

	return "", ErrNoToken
}

// GetDPoPKeyStore returns the DPoP key store.
func (am *AuthManager) GetDPoPKeyStore() *DPoPKeyStore {
	return am.keyStore
}

// GenerateDPoPProof generates a DPoP proof for the given method and URL.
//
// Parameters:
//   - method: The HTTP method
//   - url: The request URL
//
// Returns:
//   - The DPoP proof JWT
//   - Error if proof generation fails
func (am *AuthManager) GenerateDPoPProof(method, url string) (string, error) {
	// Get current key
	keyID := am.currentKeyID
	if keyID == "" {
		// Try to get current key from store
		currentKey, err := am.keyStore.GetCurrentKey()
		if err != nil {
			// Generate a new key
			key, err := am.keyStore.GenerateKey(RS256, "")
			if err != nil {
				return "", err
			}
			keyID = key.ID
			am.currentKeyID = keyID
		} else {
			keyID = currentKey.ID
			am.currentKeyID = keyID
		}
	}

	// Get access token
	accessToken, err := am.GetAccessToken()
	if err != nil {
		return "", err
	}

	return am.keyStore.GenerateProof(keyID, method, url, accessToken)
}

// GetDPoPProofFunc returns a function that generates DPoP proofs.
//
// Returns:
//   - A function that generates DPoP proofs
func (am *AuthManager) GetDPoPProofFunc() func(method, url string) (string, error) {
	return func(method, url string) (string, error) {
		return am.GenerateDPoPProof(method, url)
	}
}

// DiscoverIssuer discovers the OAuth2 issuer configuration.
//
// Parameters:
//   - ctx: Context for cancellation/timeout
//
// Returns:
//   - The issuer configuration
//   - Error if discovery fails
func (am *AuthManager) DiscoverIssuer(ctx context.Context) (*IssuerConfig, error) {
	// Build issuer metadata URL
	issuerURL, err := url.Parse(am.issuer)
	if err != nil {
		return nil, err
	}

	// Add .well-known/openid-configuration
	issuerURL.Path = ".well-known/openid-configuration"

	// Create request
	req, err := http.NewRequestWithContext(ctx, "GET", issuerURL.String(), nil)
	if err != nil {
		return nil, err
	}

	// Set headers
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "SolidSidecar-Go-SDK/1.0.0")

	// Execute request
	resp, err := am.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	// Check status
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("issuer discovery failed with status: %d", resp.StatusCode)
	}

	// Parse response
	var config IssuerConfig
	if err := json.NewDecoder(resp.Body).Decode(&config); err != nil {
		return nil, err
	}

	// Validate required fields
	if config.Issuer == "" {
		return nil, ErrInvalidIssuer
	}

	return &config, nil
}

// ExchangeCode exchanges an authorization code for an access token.
//
// Parameters:
//   - ctx: Context for cancellation/timeout
//   - code: The authorization code
//   - verifier: The PKCE code verifier
//
// Returns:
//   - The token response
//   - Error if exchange fails
func (am *AuthManager) ExchangeCode(ctx context.Context, code, verifier string) (*types.TokenResponse, error) {
	// Discover issuer config
	config, err := am.DiscoverIssuer(ctx)
	if err != nil {
		return nil, err
	}

	if config.TokenEndpoint == "" {
		return nil, ErrInvalidIssuer
	}

	// Prepare request body
	body := url.Values{}
	body.Set("grant_type", "authorization_code")
	body.Set("code", code)
	body.Set("redirect_uri", am.redirectURI)
	body.Set("client_id", am.clientID)

	if am.clientSecret != "" {
		body.Set("client_secret", am.clientSecret)
	}

	if verifier != "" {
		body.Set("code_verifier", verifier)
	}

	// Create request
	req, err := http.NewRequestWithContext(ctx, "POST", config.TokenEndpoint, strings.NewReader(body.Encode()))
	if err != nil {
		return nil, err
	}

	// Set headers
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "SolidSidecar-Go-SDK/1.0.0")

	// Execute request
	resp, err := am.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	// Check status
	if resp.StatusCode != 200 {
		// Try to parse error response
		var errResp struct {
			Error            string `json:"error"`
			ErrorDescription string `json:"error_description,omitempty"`
		}
		if err := json.NewDecoder(resp.Body).Decode(&errResp); err == nil {
			if errResp.Error != "" {
				return nil, fmt.Errorf("token exchange failed: %s - %s", errResp.Error, errResp.ErrorDescription)
			}
		}
		return nil, fmt.Errorf("token exchange failed with status: %d", resp.StatusCode)
	}

	// Parse response
	var tokenResp types.TokenResponse
	if err := json.NewDecoder(resp.Body).Decode(&tokenResp); err != nil {
		return nil, err
	}

	// Set token in manager
	am.tokenManager.SetToken(&tokenResp, tokenResp.ExpiresIn)

	// Set issuer
	am.tokenManager.SetIssuer(am.issuer)
	am.tokenManager.SetClientID(am.clientID)
	am.tokenManager.SetScope(am.scope)

	return &tokenResp, nil
}

// RefreshToken refreshes the access token.
//
// Returns:
//   - Error if refresh fails
func (am *AuthManager) refreshToken() error {
	// Discover issuer config
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	config, err := am.DiscoverIssuer(ctx)
	if err != nil {
		return err
	}

	if config.TokenEndpoint == "" {
		return ErrInvalidIssuer
	}

	// Get current refresh token
	refreshToken := am.tokenManager.GetRefreshToken()
	if refreshToken == "" {
		return errors.New("no refresh token available")
	}

	// Prepare request body
	body := url.Values{}
	body.Set("grant_type", "refresh_token")
	body.Set("refresh_token", refreshToken)
	body.Set("client_id", am.clientID)

	if am.clientSecret != "" {
		body.Set("client_secret", am.clientSecret)
	}

	// Create request
	req, err := http.NewRequestWithContext(ctx, "POST", config.TokenEndpoint, strings.NewReader(body.Encode()))
	if err != nil {
		return err
	}

	// Set headers
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "SolidSidecar-Go-SDK/1.0.0")

	// Execute request
	resp, err := am.httpClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	// Check status
	if resp.StatusCode != 200 {
		// Try to parse error response
		var errResp struct {
			Error            string `json:"error"`
			ErrorDescription string `json:"error_description,omitempty"`
		}
		if err := json.NewDecoder(resp.Body).Decode(&errResp); err == nil {
			if errResp.Error != "" {
				return fmt.Errorf("token refresh failed: %s - %s", errResp.Error, errResp.ErrorDescription)
			}
		}
		return fmt.Errorf("token refresh failed with status: %d", resp.StatusCode)
	}

	// Parse response
	var tokenResp types.TokenResponse
	if err := json.NewDecoder(resp.Body).Decode(&tokenResp); err != nil {
		return err
	}

	// Set token in manager
	am.tokenManager.SetToken(&tokenResp, tokenResp.ExpiresIn)

	// Call refresh callback if set
	if am.onTokenRefresh != nil {
		am.onTokenRefresh(tokenResp.AccessToken)
	}

	return nil
}

// RevokeToken revokes an access token.
//
// Parameters:
//   - ctx: Context for cancellation/timeout
//   - token: The token to revoke (if empty, uses current access token)
//
// Returns:
//   - Error if revocation fails
func (am *AuthManager) RevokeToken(ctx context.Context, token string) error {
	if token == "" {
		var err error
		token, err = am.GetAccessToken()
		if err != nil {
			return err
		}
	}

	// Discover issuer config
	config, err := am.DiscoverIssuer(ctx)
	if err != nil {
		return err
	}

	if config.RevocationEndpoint == "" {
		// If no revocation endpoint, just clear local token
		am.tokenManager.Clear()
		return nil
	}

	// Prepare request body
	body := url.Values{}
	body.Set("token", token)
	body.Set("client_id", am.clientID)

	if am.clientSecret != "" {
		body.Set("client_secret", am.clientSecret)
	}

	// Create request
	req, err := http.NewRequestWithContext(ctx, "POST", config.RevocationEndpoint, strings.NewReader(body.Encode()))
	if err != nil {
		return err
	}

	// Set headers
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")
	req.Header.Set("User-Agent", "SolidSidecar-Go-SDK/1.0.0")

	// Execute request
	resp, err := am.httpClient.Do(req)
	if err != nil {
		// Clear local token even if server call fails
		am.tokenManager.Clear()
		return err
	}
	defer resp.Body.Close()

	// Check status
	if resp.StatusCode != 200 {
		// Clear local token
		am.tokenManager.Clear()
		return fmt.Errorf("token revocation failed with status: %d", resp.StatusCode)
	}

	// Clear local token
	am.tokenManager.Clear()
	return nil
}

// ClearSession clears the current session.
func (am *AuthManager) ClearSession() {
	am.tokenManager.Clear()
	am.currentKeyID = ""
}

// SetOnTokenRefresh sets the callback for token refresh.
//
// Parameters:
//   - callback: The callback function
func (am *AuthManager) SetOnTokenRefresh(callback func(token string)) {
	am.onTokenRefresh = callback
}

// EnableAutoRefresh enables automatic token refresh.
//
// Parameters:
//   - enabled: Whether to enable auto-refresh
func (am *AuthManager) EnableAutoRefresh(enabled bool) {
	am.autoRefresh = enabled
}

// IssuerConfig represents OAuth2 issuer configuration.
type IssuerConfig struct {
	// Issuer is the issuer URL
	Issuer string `json:"issuer"`

	// AuthorizationEndpoint is the authorization endpoint
	AuthorizationEndpoint string `json:"authorization_endpoint"`

	// TokenEndpoint is the token endpoint
	TokenEndpoint string `json:"token_endpoint"`

	// RevocationEndpoint is the revocation endpoint (optional)
	RevocationEndpoint string `json:"revocation_endpoint,omitempty"`

	// JWKSURI is the JWKS endpoint
	JWKSURI string `json:"jwks_uri"`

	// ScopesSupported are the supported scopes
	ScopesSupported []string `json:"scopes_supported"`

	// ResponseTypesSupported are the supported response types
	ResponseTypesSupported []string `json:"response_types_supported"`

	// SubjectTypesSupported are the supported subject types
	SubjectTypesSupported []string `json:"subject_types_supported"`

	// IDTokenSigningAlgValuesSupported are the supported signing algorithms
	IDTokenSigningAlgValuesSupported []string `json:"id_token_signing_alg_values_supported"`

	// DPoPSigningAlgValuesSupported are the supported DPoP signing algorithms
	DPoPSigningAlgValuesSupported []string `json:"dpop_signing_alg_values_supported,omitempty"`
}

// ValidateDPoPSupport checks if the issuer supports DPoP.
//
// Parameters:
//   - config: The issuer configuration
//
// Returns:
//   - true if DPoP is supported
func ValidateDPoPSupport(config *IssuerConfig) bool {
	// Check for DPoP-specific endpoints or algorithms
	if config == nil {
		return false
	}

	// Check if DPoP signing algorithms are specified
	if len(config.DPoPSigningAlgValuesSupported) > 0 {
		return true
	}

	// Check common DPoP signing algorithms in regular signing algs
	for _, alg := range config.IDTokenSigningAlgValuesSupported {
		switch alg {
		case string(RS256), string(ES256), string(ES384), string(ES512), string(EdDSA):
			return true
		}
	}

	return false
}

// import "crypto/tls" is needed for the AuthManager

File: sdk/ts/src/auth/auth-manager.ts
/**
 * Solid Sidecar TypeScript SDK - Authentication Manager
 * Phase: 27 - SDK/Client Compatibility Layer
 * License: MIT
 */

import type {
  DpopProof,
  DpopProofOptions,
  PkcePair,
  SolidSidecarLogger,
  TokenResponse,
  TokenSet,
} from '../types';
import { AuthenticationError, ValidationError } from '../types';
import {
  DEFAULT_DPOP_KEYSTORE_CONFIG,
  DpopKeyStore,
  type DpopKeyStoreConfig,
} from './dpop-keystore';
import {
  base64UrlEncode,
  generateUuid,
  nullLogger,
  sha256Base64UrlSync,
} from '../utils';

export interface AuthManagerConfig {
  issuerUrl: string;
  clientId: string;
  clientSecret?: string;
  redirectUri: string;
  scopes?: string[];
  dpopKeyStoreConfig?: DpopKeyStoreConfig;
  useDpop?: boolean;
  logger?: SolidSidecarLogger;
}

type ResolvedAuthManagerConfig = Omit<
  Required<AuthManagerConfig>,
  'clientSecret'
> & {
  clientSecret?: string;
};

export const DEFAULT_AUTH_MANAGER_CONFIG: Pick<
  ResolvedAuthManagerConfig,
  'scopes' | 'useDpop' | 'dpopKeyStoreConfig' | 'logger'
> = {
  scopes: ['openid', 'profile', 'webid'],
  useDpop: true,
  dpopKeyStoreConfig: DEFAULT_DPOP_KEYSTORE_CONFIG,
  logger: nullLogger,
};

export interface TokenStorage {
  getTokenSet(): Promise<TokenSet | null>;
  setTokenSet(tokenSet: TokenSet): Promise<void>;
  deleteTokenSet(): Promise<void>;
  getAccessToken(): Promise<string | null>;
  getRefreshToken(): Promise<string | null>;
  getTokenExpiration(): Promise<number | null>;
}

export class InMemoryTokenStorage implements TokenStorage {
  private tokenSet: TokenSet | null = null;

  public async getTokenSet(): Promise<TokenSet | null> {
    return this.tokenSet;
  }

  public async setTokenSet(tokenSet: TokenSet): Promise<void> {
    this.tokenSet = tokenSet;
  }

  public async deleteTokenSet(): Promise<void> {
    this.tokenSet = null;
  }

  public async getAccessToken(): Promise<string | null> {
    return this.tokenSet?.access_token || null;
  }

  public async getRefreshToken(): Promise<string | null> {
    return this.tokenSet?.refresh_token || null;
  }

  public async getTokenExpiration(): Promise<number | null> {
    return this.tokenSet?.expires_at || null;
  }
}

/**
 * In-memory encrypted token storage for Node.js processes.
 *
 * This protects token material from accidental plaintext exposure inside this
 * storage object, but it is not durable storage and is not a replacement for
 * an operating-system or hardware-backed secret store.
 */
export class SecureTokenStorage implements TokenStorage {
  private readonly storage = new InMemoryTokenStorage();
  private readonly encryptionKey: Uint8Array;

  constructor(encryptionKey?: string) {
    if (encryptionKey) {
      this.encryptionKey = new TextEncoder().encode(encryptionKey);
    } else {
      this.encryptionKey = require('crypto').randomBytes(32);
    }
  }

  private async encrypt(data: TokenSet): Promise<TokenSet> {
    const crypto = require('crypto');
    const iv = crypto.randomBytes(16);
    const cipher = crypto.createCipheriv(
      'aes-256-gcm',
      this.encryptionKey,
      iv
    );
    const encrypted = Buffer.concat([
      cipher.update(JSON.stringify(data), 'utf8'),
      cipher.final(),
      cipher.getAuthTag(),
    ]);

    return {
      ...data,
      access_token: Buffer.concat([iv, encrypted]).toString('base64'),
      refresh_token: '',
    };
  }

  private async decrypt(encrypted: TokenSet): Promise<TokenSet> {
    const crypto = require('crypto');
    const data = Buffer.from(encrypted.access_token || '', 'base64');
    const iv = data.subarray(0, 16);
    const ciphertext = data.subarray(16, data.length - 16);
    const authTag = data.subarray(data.length - 16);
    const decipher = crypto.createDecipheriv(
      'aes-256-gcm',
      this.encryptionKey,
      iv
    );
    decipher.setAuthTag(authTag);
    const decrypted = Buffer.concat([
      decipher.update(ciphertext),
      decipher.final(),
    ]).toString('utf8');

    return JSON.parse(decrypted) as TokenSet;
  }

  public async getTokenSet(): Promise<TokenSet | null> {
    const encrypted = await this.storage.getTokenSet();
    return encrypted ? this.decrypt(encrypted) : null;
  }

  public async setTokenSet(tokenSet: TokenSet): Promise<void> {
    await this.storage.setTokenSet(await this.encrypt(tokenSet));
  }

  public async deleteTokenSet(): Promise<void> {
    await this.storage.deleteTokenSet();
  }

  public async getAccessToken(): Promise<string | null> {
    return (await this.getTokenSet())?.access_token || null;
  }

  public async getRefreshToken(): Promise<string | null> {
    return (await this.getTokenSet())?.refresh_token || null;
  }

  public async getTokenExpiration(): Promise<number | null> {
    return (await this.getTokenSet())?.expires_at || null;
  }
}

export function generatePkcePair(): PkcePair {
  const codeVerifier = base64UrlEncode(require('crypto').randomBytes(32));
  return {
    codeVerifier,
    codeChallenge: sha256Base64UrlSync(codeVerifier),
    codeChallengeMethod: 'S256',
  };
}

export class AuthManager {
  private readonly config: ResolvedAuthManagerConfig;
  private readonly dpopKeyStore: DpopKeyStore;
  private readonly tokenStorage: TokenStorage;
  private readonly logger: SolidSidecarLogger;

  constructor(config: AuthManagerConfig) {
    if (!config.issuerUrl) {
      throw new ValidationError('OIDC issuer URL is required', 'issuerUrl');
    }
    if (!config.clientId) {
      throw new ValidationError('Client ID is required', 'clientId');
    }
    if (!config.redirectUri) {
      throw new ValidationError('Redirect URI is required', 'redirectUri');
    }

    this.config = {
      ...DEFAULT_AUTH_MANAGER_CONFIG,
      ...config,
      scopes: config.scopes ?? DEFAULT_AUTH_MANAGER_CONFIG.scopes,
      useDpop: config.useDpop ?? DEFAULT_AUTH_MANAGER_CONFIG.useDpop,
      dpopKeyStoreConfig:
        config.dpopKeyStoreConfig ??
        DEFAULT_AUTH_MANAGER_CONFIG.dpopKeyStoreConfig,
      logger: config.logger ?? DEFAULT_AUTH_MANAGER_CONFIG.logger,
    };

    this.dpopKeyStore = new DpopKeyStore(this.config.dpopKeyStoreConfig);
    this.tokenStorage = new InMemoryTokenStorage();
    this.logger = this.config.logger;

    this.logger.debug('AuthManager initialized', {
      issuerUrl: this.config.issuerUrl,
      clientId: this.config.clientId,
      redirectUri: this.config.redirectUri,
      useDpop: this.config.useDpop,
    });
  }

  public generatePkce(): PkcePair {
    return generatePkcePair();
  }

  public buildAuthorizationUrl(
    pkce: PkcePair,
    state?: string,
    nonce?: string
  ): string {
    const { issuerUrl, clientId, redirectUri, scopes } = this.config;
    const url = new URL(issuerUrl);
    url.pathname = `${url.pathname.replace(/\/$/, '')}/auth`;
    url.search = new URLSearchParams({
      response_type: 'code',
      client_id: clientId,
      redirect_uri: redirectUri,
      code_challenge: pkce.codeChallenge,
      code_challenge_method: pkce.codeChallengeMethod,
      scope: scopes.join(' '),
      state: state || generateUuid(),
      nonce: nonce || generateUuid(),
    }).toString();
    return url.toString();
  }

  public async exchangeCode(
    code: string,
    codeVerifier: string
  ): Promise<TokenSet> {
    if (!code) {
      throw new AuthenticationError('Authorization code is required');
    }
    if (!codeVerifier) {
      throw new AuthenticationError('PKCE code verifier is required');
    }

    const { issuerUrl, clientId, redirectUri, clientSecret } = this.config;
    const url = new URL(issuerUrl);
    url.pathname = `${url.pathname.replace(/\/$/, '')}/token`;
    const params = new URLSearchParams({
      grant_type: 'authorization_code',
      code,
      redirect_uri: redirectUri,
      client_id: clientId,
      code_verifier: codeVerifier,
    });
    if (clientSecret) {
      params.set('client_secret', clientSecret);
    }

    const response = await this.fetchWithRetry(url.toString(), {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: params.toString(),
    });
    if (!response.ok) {
      const error = await response.text();
      this.logger.error('Token exchange failed', {
        status: response.status,
        error,
      });
      throw new AuthenticationError(
        `Token exchange failed: ${response.status}`,
        { statusCode: response.status }
      );
    }

    const tokenResponse: TokenResponse = await response.json();
    const now = Date.now();
    const tokenSet: TokenSet = {
      ...tokenResponse,
      expires_at: now + tokenResponse.expires_in * 1000,
      issued_at: tokenResponse.issued_at
        ? tokenResponse.issued_at * 1000
        : now,
    };
    await this.tokenStorage.setTokenSet(tokenSet);

    if (this.config.useDpop) {
      await this.dpopKeyStore.generateKeyPair(
        this.config.dpopKeyStoreConfig.defaultAlgorithm
      );
    }

    this.logger.info('Token exchange successful', {
      expires_at: tokenSet.expires_at,
      scopes: tokenResponse.scope,
    });
    return tokenSet;
  }

  public async refreshToken(): Promise<TokenSet> {
    const existingRefreshToken = await this.tokenStorage.getRefreshToken();
    if (!existingRefreshToken) {
      throw new AuthenticationError('No refresh token available');
    }

    const { issuerUrl, clientId, clientSecret } = this.config;
    const url = new URL(issuerUrl);
    url.pathname = `${url.pathname.replace(/\/$/, '')}/token`;
    const params = new URLSearchParams({
      grant_type: 'refresh_token',
      refresh_token: existingRefreshToken,
      client_id: clientId,
    });
    if (clientSecret) {
      params.set('client_secret', clientSecret);
    }

    const response = await this.fetchWithRetry(url.toString(), {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: params.toString(),
    });
    if (!response.ok) {
      const error = await response.text();
      this.logger.error('Token refresh failed', {
        status: response.status,
        error,
      });
      await this.tokenStorage.deleteTokenSet();
      throw new AuthenticationError(
        `Token refresh failed: ${response.status}`,
        { statusCode: response.status }
      );
    }

    const tokenResponse: TokenResponse = await response.json();
    const now = Date.now();
    const tokenSet: TokenSet = {
      ...tokenResponse,
      refresh_token: tokenResponse.refresh_token || existingRefreshToken,
      expires_at: now + tokenResponse.expires_in * 1000,
      issued_at: tokenResponse.issued_at
        ? tokenResponse.issued_at * 1000
        : now,
    };
    await this.tokenStorage.setTokenSet(tokenSet);
    this.logger.info('Token refresh successful', {
      expires_at: tokenSet.expires_at,
    });
    return tokenSet;
  }

  public async getAccessToken(): Promise<string | null> {
    const tokenSet = await this.tokenStorage.getTokenSet();
    if (!tokenSet) {
      return null;
    }

    const expiresAt = tokenSet.expires_at || 0;
    if (Date.now() >= expiresAt - 5 * 60 * 1000) {
      try {
        await this.refreshToken();
        return this.getAccessToken();
      } catch (error) {
        this.logger.error('Auto-refresh failed', { error });
        return null;
      }
    }
    return tokenSet.access_token;
  }

  public async getRefreshToken(): Promise<string | null> {
    return this.tokenStorage.getRefreshToken();
  }

  public async getTokenExpiration(): Promise<number | null> {
    return this.tokenStorage.getTokenExpiration();
  }

  public async clearTokens(): Promise<void> {
    await this.tokenStorage.deleteTokenSet();
    this.logger.info('Tokens cleared');
  }

  public async generateDpopProof(
    options: DpopProofOptions
  ): Promise<DpopProof> {
    if (!this.config.useDpop) {
      throw new Error(
        'DPoP is disabled. Enable DPoP in AuthManager configuration.'
      );
    }
    const accessToken = await this.getAccessToken();
    if (!accessToken) {
      throw new AuthenticationError('No access token available');
    }
    return this.dpopKeyStore.generateDpopProof({
      ...options,
      accessToken,
    });
  }

  public async signRequest(
    method: string,
    url: string
  ): Promise<{ accessToken: string; dpopProof: string }> {
    const accessToken = await this.getAccessToken();
    if (!accessToken) {
      throw new AuthenticationError('No access token available');
    }
    const dpopProof = await this.generateDpopProof({
      method: method as any,
      url,
      accessToken,
    });
    return { accessToken, dpopProof: dpopProof.jwt };
  }

  private async fetchWithRetry(
    url: string,
    options: RequestInit,
    retries = 3
  ): Promise<Response> {
    try {
      const fetchImpl =
        typeof fetch === 'undefined' ? require('cross-fetch') : fetch;
      return await fetchImpl(url, options);
    } catch (error) {
      if (retries <= 0) {
        throw error;
      }
      const delay = Math.pow(2, 3 - retries) * 1000;
      await new Promise(resolve => setTimeout(resolve, delay));
      return this.fetchWithRetry(url, options, retries - 1);
    }
  }

  public getDpopKeyStore(): DpopKeyStore {
    return this.dpopKeyStore;
  }

  public getTokenStorage(): TokenStorage {
    return this.tokenStorage;
  }

  public async isAuthenticated(): Promise<boolean> {
    return !!(await this.getAccessToken());
  }

  public async getSessionInfo(): Promise<{
    isAuthenticated: boolean;
    accessToken: string | null;
    expiresAt: number | null;
    clientId: string;
    issuerUrl: string;
    useDpop: boolean;
  }> {
    const [accessToken, expiresAt] = await Promise.all([
      this.getAccessToken(),
      this.getTokenExpiration(),
    ]);
    return {
      isAuthenticated: !!accessToken,
      accessToken: accessToken ? '[REDACTED]' : null,
      expiresAt,
      clientId: this.config.clientId,
      issuerUrl: this.config.issuerUrl,
      useDpop: this.config.useDpop,
    };
  }
}
